tests/unit/test_pillar.py

```diff
@@ -895,3 +896,42 @@ def test_pillar_send_extra_minion_data_from_config(self):
              'pillar_override': {},
              'extra_minion_data': {'path_to_add': 'fake_data'}},
             dictkey='pillar')
+
+
+@skipIf(NO_MOCK, NO_MOCK_REASON)
+class PillarCacheTestCase(TestCase):
+    '''
+    Tests for instantiating a PillarCache in salt.pillar
+    '''
+    def setUp(self):
+        self.tmpdir = tempfile.mkdtemp()
+        self.path = os.path.join(self.tmpdir, 'CacheDisk_test')
+
+    def tearDown(self):
+        shutil.rmtree(self.tmpdir, ignore_errors=True)
+
+    def test_compile_pillar(self):
+        opts = {
+            'pillar_cache': True,
+            'pillar_cache_ttl': 10,
+            'pillar_cache_backend': 'disk'
+        }
+
+        with patch('salt.pillar.PillarCache._minion_cache_path', MagicMock(return_value=self.path)):
+            # We haven't seen this minion yet in the cache. Store it.
+            pillar = salt.pillar.PillarCache(opts, {}, 'mocked-minion', 'dev', pillarenv='dev')
+            with patch('salt.pillar.PillarCache.fetch_pillar', MagicMock(return_value={'env': 'dev'})):
+                pillar.compile_pillar()
+                self.assertEqual(pillar.cache['mocked-minion'], {'dev': {'env': 'dev'}})
+            # We found the minion but not the env. Store it.
+            pillar = salt.pillar.PillarCache(opts, {}, 'mocked-minion', 'test', pillarenv='test')
+            with patch('salt.pillar.PillarCache.fetch_pillar', MagicMock(return_value={'env': 'test'})):
+                pillar.compile_pillar()
+                self.assertEqual(pillar.cache['mocked-minion'], {'dev': {'env': 'dev'},
+                                                                 'test': {'env': 'test'}})
+            # We have a cache hit! Send it back.
+            pillar = salt.pillar.PillarCache(opts, {}, 'mocked-minion', 'test', pillarenv='test')
+            with patch('salt.pillar.PillarCache.fetch_pillar', MagicMock(return_value={'k': 'v'})):
+                pillar.compile_pillar()
+                self.assertEqual(pillar.cache['mocked-minion'], {'dev': {'env': 'dev'},
+                                                                 'test': {'env': 'test'}})
```

Comment:
The last test doesn't make sense. You should actually explicitly modify time so TTL changes and then you get back `k/v` instead if `env/test`. And you should explicitly hold time so the cache works.

But you continue developing this, **directly ignoring** requests to explain why this is even needed and what you're trying to solve here.

Reply:
Hi @isbm 
The last test is for testing whether the cache was actually hit but not for testing TTL, because the TTL has been tested in file https://github.com/saltstack/salt/blob/develop/tests/unit/utils/test_cache.py

And I think i have tried to explained why this helps, and what problem this fixes. 
```
self.cache[self.minion_id][self.pillarenv] = x
Will only call the __getitem__ of CacheDisk class, and will not call the __setitem__

self.cache[self.minion_id] = x
Will only call the __setitem__ of CacheDisk class
```

In here, the second is correct.

Maybe you should run the test first.

Reply:
@rongzeng54 but it should not. I am not sure you understand how this supposed to work. Here is an example:

```python
from salt.utils.cache import CacheDisk
import time
import os

cache_file = '/tmp/spam_eggs'
cache = CacheDisk(0, cache_file)

if not os.path.exists(cache_file):
    cache['foo'] = {}
    cache['foo']['bar'] = {}
    cache['foo']['bar']['spam'] = 'Fred'
    cache._write()
else:
    cache._read()
    print(cache['foo'])
    cache['foo'][str(time.time())] = 'one more time'
    cache._write()
```

So first time when you call this, it will  do nothing:
```
$ python cd.py
$
```

But then each time you call it, it will add more data:
```
$ python cd.py 
{'bar': {'spam': 'Fred'}}
$ python cd.py 
{'bar': {'spam': 'Fred'}, '1542639980.9734464': 'one more time'}
$ python cd.py 
{'bar': {'spam': 'Fred'}, '1542639980.9734464': 'one more time', '1542639982.2767625': 'one more time'}
```

What part is still missing you think?

Reply:
@isbm it should call __setitem__. Because the function _write used for Write out the pillar data to disk exists in __setitem__ but not __getitem__.